mlir/docs/Tutorials/DataFlowAnalysis.md

```diff
@@ -5,20 +5,361 @@ daunting and/or complex. A dataflow analysis generally involves propagating
+    // Try to compute a constant value of the result.
+    auto *result = getOrCreate<StringConstant>(op->getResult(0));
+    if (auto constant = dyn_cast<string::ConstantOp>(op)) {
+      // Just grab and set the constant value of the result of the operation.
+      // Propagate an update to the state if it changed.
+      propagateIfChanged(result, result->join(constant.getValue()));
+    } else if (auto concat = dyn_cast<string::ConcatOp>(op)) {
+      StringRef lhs = operandValues[0]->getStringValue();
+      StringRef rhs = operandValues[1]->getStringValue();
+      // If either operand is overdefined, the results are overdefined.
+      if (lhs.empty() || rhs.empty()) {
+        propagateIfChanged(result, result->defaultInitialize());
+
+        // Otherwise, compute the constant value and join it with the result.
+      } else {
+        propagateIfChanged(result, result->join(lhs + rhs));
+      }
+    } else {
+      // We don't know how to implement the transfer function for this
+      // operation. Mark its results as overdefined.
+      propagateIfChanged(result, result->defaultInitialize());
+    }
+    return success();
+  }
+};
+```
+
+In the above example, the `visit` function sets up the dependencies of the
+analysis invocation on an operation as the constant values of the operands of
+each operation. When the operand states have initialized values but overdefined
+values, it sets the state of the result to overdefined. Otherwise, it computes
+the state of the result and merges the new information in with `join`.
+
+However, the dependency graph still needs to be initialized before the solver
+knows what to call `visit` on. This is done in the `initialize` function:
+
+```c++
+LogicalResult StringConstantPropagation::initialize(Operation *top) {
+  // Visit every nested string operation and set up its dependencies.
+  top->walk([&](Operation *op) {
+    for (Value operand : op->getOperands()) {
+      auto *state = getOrCreate<StringConstant>(operand);
+      addDependency(state, getProgramPointAfter(op));
+    }
+  });
+  // Now that the dependency graph has been set up, "seed" the evolution of the
+  // analysis by marking the constant values of all block arguments as
+  // overdefined and the results of (non-constant) operations with no operands.
+  auto defaultInitializeAll = [&](ValueRange values) {
+    for (Value value : values) {
+      auto *state = getOrCreate<StringConstant>(value);
+      propagateIfChanged(state, state->defaultInitialize());
+    }
+  };
+  top->walk([&](Operation *op) {
+    for (Region &region : op->getRegions())
+      for (Block &block : region)
+        defaultInitializeAll(block.getArguments());
+    if (auto constant = dyn_cast<string::ConstantOp>(op)) {
+      auto *result = getOrCreate<StringConstant>(constant.getResult());
+      propagateIfChanged(result, result->join(constant.getValue()));
+    } else if (op->getNumOperands() == 0) {
+      defaultInitializeAll(op->getResults());
+    }
```

Comment:
I don't think this is a good initialization. With this setup, the analysis will never find any constant values for block arguments or across control flow, which it is otherwise perfectly capable of...